core/src/main/java/hudson/security/HudsonPrivateSecurityRealm.java

```diff
@@ -804,56 +860,7 @@ public Category getCategory() {
         }
     }
 
-    /**
-     * {@link PasswordEncoder} based on SHA-256 and random salt generation.
-     *
-     * <p>
-     * The salt is prepended to the hashed password and returned. So the encoded password is of the form
-     * {@code SALT ':' hash(PASSWORD,SALT)}.
-     *
-     * <p>
-     * This abbreviates the need to store the salt separately, which in turn allows us to hide the salt handling
-     * in this little class. The rest of the Acegi thinks that we are not using salt.
-     */
-    /*package*/ static final PasswordEncoder CLASSIC = new PasswordEncoder() {
-        private final PasswordEncoder passwordEncoder = new ShaPasswordEncoder(256);
-
-        public String encodePassword(String rawPass, Object obj) throws DataAccessException {
-            return hash(rawPass);
-        }
-
-        public boolean isPasswordValid(String encPass, String rawPass, Object obj) throws DataAccessException {
-            // pull out the sale from the encoded password
-            int i = encPass.indexOf(':');
-            if(i<0) return false;
-            String salt = encPass.substring(0,i);
-            return encPass.substring(i+1).equals(passwordEncoder.encodePassword(rawPass,salt));
-        }
-
-        /**
-         * Creates a hashed password by generating a random salt.
-         */
-        private String hash(String password) {
-            String salt = generateSalt();
-            return salt+':'+passwordEncoder.encodePassword(password,salt);
-        }
-
-        /**
-         * Generates random salt.
-         */
-        private String generateSalt() {
-            StringBuilder buf = new StringBuilder();
-            SecureRandom sr = new SecureRandom();
-            for( int i=0; i<6; i++ ) {// log2(52^6)=34.20... so, this is about 32bit strong.
-                boolean upper = sr.nextBoolean();
-                char ch = (char)(sr.nextInt(26) + 'a');
-                if(upper)   ch=Character.toUpperCase(ch);
-                buf.append(ch);
-            }
-            return buf.toString();
-        }
-    };
-
+    // TODO can we instead use BCryptPasswordEncoder from Spring Security, which has its own copy of BCrypt so we could drop the special library?
```

Comment:
We cannot use BouncyCastle from core, though.

Reply:
I suppose it shouldn't matter for this case, then, since Spring Security already includes a BCrypt implementation in its own dependencies like you mentioned in the comment here. Does this limitation include for remoting? That'd be annoying if the only way to support TLS 1.3 connections there is to use Java 11 or customizing your classpath manually.

Reply:
Not sure what you are talking about. BouncyCastle is bundled only in a wrapper plugin for use by other plugins.

Reply:
BouncyCastle has JSSE glue to support TLS 1.3 in older JDKs than 11. Classes like `SSLContext` call into that which is the usual class for handling TLS code in Java.

Reply:
Sorry, do not know anything about it. Off topic I suppose. The point here is that for now I kept the existing behavior of using our own bundled jBCrypt, but now that Spring Security also bundles their own copy, we could consider switching to that and dropping our dep. I did not attempt to do so here because the format uses by `BCryptPasswordEncoder` does not match that of existing saved passwords in Jenkins, so we would need to do more work.